firebase-firestore/src/test/java/com/google/firebase/firestore/local/BundleSerializerTest.java

```diff
@@ -0,0 +1,701 @@
+    Query query =
+        TestUtil.query("coll")
+            .orderBy(orderBy("foo"))
+            .endAt(
+                new Bound(
+                    Collections.singletonList(Value.newBuilder().setStringValue("bar").build()),
+                    true));
+    assertNamedQueryRoundTrip(json, query);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testDoesNotDecodeOffset() throws JSONException {
+    String json = "{ from: [ { collectionId: 'coll' } ], offset: 5 }";
+    Query query = TestUtil.query("coll");
+    assertNamedQueryRoundTrip(json, query);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testDoesNotDecodeSelect() throws JSONException {
+    String json = "{ from: [ { collectionId: 'coll' } ], select: [] }";
+    Query query = TestUtil.query("coll");
+    assertNamedQueryRoundTrip(json, query);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testDoesNotDecodeMissingCollection() throws JSONException {
+    String json = "{ from: [ ] }";
+    Query query = TestUtil.query("coll");
+    assertNamedQueryRoundTrip(json, query);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testDoesNotDecodeMultipleCollections() throws JSONException {
+    String json = "{ from: [ { collectionId: 'c1' }, { collectionId: 'c2' } ] }";
+    Query query = TestUtil.query("coll");
+    assertNamedQueryRoundTrip(json, query);
+  }
+
+  // BundleMetadata tests
+
+  @Test
+  public void testDecodesBundleMetadata() throws JSONException {
+    String json =
+        "{\n"
+            + "id: 'bundle-1',\n"
+            + "version: 1,\n"
+            + "createTime: { seconds: 2, nanos: 3 },\n"
+            + "totalDocuments: 4,\n"
+            + "totalBytes: 5\n"
+            + "}";
+    BundleMetadata expectedMetadata =
+        new BundleMetadata(
+            "bundle-1", 1, new SnapshotVersion(new com.google.firebase.Timestamp(2, 3)), 4, 5);
+    BundleMetadata actualMetadata = serializer.fromBundleMetadata(json);
+    assertEquals(expectedMetadata, actualMetadata);
+  }
+
+  // BundledDocumentMetadata tests
+
+  @Test
+  public void testDecodesBundledDocumentMetadata() throws JSONException {
+    String json =
+        "{\n"
+            + "name: '"
+            + TEST_DOCUMENT
+            + "',\n"
+            + "readTime: { seconds: 1, nanos: 2 },\n"
+            + "exists: true,\n"
+            + "queries: [ 'query-1', 'query-2' ]\n"
+            + "}";
+    BundledDocumentMetadata expectedMetadata =
+        new BundledDocumentMetadata(
+            key("coll/doc"),
+            new SnapshotVersion(new com.google.firebase.Timestamp(1, 2)),
+            true,
+            Arrays.asList("query-1", "query-2"));
+    BundledDocumentMetadata actualMetadata = serializer.fromBundledDocumentMetadata(json);
+    assertEquals(expectedMetadata, actualMetadata);
+  }
+
+  private void assertValueRoundTrip(String json, Value proto) throws JSONException {
```

Comment:
Since this isn't actually round-tripping the values, should this be named `assertDecodes`, `assertDecodesValue`, or something like that?

Also applies to other methods with a `RoundTrip` suffix.

Reply:
Done